group_theory/sylow.lean

```diff
@@ -0,0 +1,508 @@
+      ⟨⟨1, is_submonoid.one_mem H⟩⟩)).1
+    (by rwa [← card_eq_card_cosets_mul_card_subgroup, hH2, hs,
+      nat.pow_succ, mul_assoc, mul_comm p]),
+have hm : s * p % p = card (left_cosets {x : normalizer H | ↑x ∈ H}) % p :=
+  card_congr (fixed_points_mul_left_cosets_equiv_left_cosets H) ▸ hcard ▸
+    card_modeq_card_fixed_points _ hp hH2,
+have hm' : p ∣ card (left_cosets {x : normalizer H | ↑x ∈ H}) :=
+  nat.dvd_of_mod_eq_zero
+    (by rwa [nat.mod_eq_zero_of_dvd (dvd_mul_left _ _), eq_comm] at hm),
+let ⟨x, hx⟩ := exists_prime_order_of_dvd_card hp hm' in
+have hxcard : ∀ {f : fintype (gpowers x)}, card (gpowers x) = p,
+  from λ f, by rw [← hx, order_eq_card_gpowers]; congr,
+have is_subgroup (left_cosets.mk ⁻¹' gpowers x),
+  from is_group_hom.preimage _ _,
+have fintype (left_cosets.mk ⁻¹' gpowers x), by apply_instance,
+have hequiv : H ≃ {x : normalizer H | ↑x ∈ H }:=
+  ⟨λ a, ⟨⟨a.1, subset_normalizer _ a.2⟩, a.2⟩, λ a, ⟨a.1.1, a.2⟩,
+    λ ⟨_, _⟩, rfl, λ ⟨⟨_, _⟩, _⟩, rfl⟩,
+⟨subtype.val '' (left_cosets.mk ⁻¹' gpowers x), by apply_instance,
+  by rw [set.card_image_of_injective _ subtype.val_injective, nat.pow_succ,
+    card_preimage_left_cosets_mk, hxcard, ← hH2, card_congr hequiv]; congr⟩
+
+def conjugate_set (x : G) (H : set G) : set G :=
+(λ n, x⁻¹ * n * x) ⁻¹' H
+
+lemma conjugate_set_eq_image (H : set G) (x : G) :
+  conjugate_set x H = (λ n, x * n * x⁻¹) '' H :=
+eq.symm (congr_fun (set.image_eq_preimage_of_inverse
+  (λ _, by simp [mul_assoc]) (λ _, by simp [mul_assoc])) _)
+
+lemma conjugate_set_eq_preimage (H : set G) (x : G) :
+  conjugate_set x H = (λ n, x⁻¹ * n * x) ⁻¹' H := rfl
+
+instance conjugate_set.is_group_action : is_group_action (@conjugate_set G _) :=
+{ one := λ H, by simp [conjugate_set_eq_image, set.image],
+  mul := λ x y H, by simp [mul_inv_rev, mul_assoc, function.comp, conjugate_set_eq_preimage,
+    set.preimage] }
+
+@[simp] lemma conjugate_set_normal_subgroup (H : set G) [normal_subgroup H] (x : G) :
+  conjugate_set x H = H :=
+set.ext (λ n, ⟨λ h : _ ∈ H, by have := normal_subgroup.normal _ h x; simpa [mul_assoc] using this,
+λ h, show _ ∈ H, by have := normal_subgroup.normal _ h (x⁻¹); by simpa using this⟩)
+
+instance is_group_action.subgroup (H : set G) [is_subgroup H] (f : G → α → α) [is_group_action f] :
+  is_group_action (λ x : H, f x) :=
+{ one := λ a, is_monoid_action.one f a,
+  mul := λ ⟨x, hx⟩ ⟨y, hy⟩ a, is_monoid_action.mul f x y a }
+
+instance is_group_hom_conj (x : G) : is_group_hom (λ (n : G), x * n * x⁻¹) :=
+⟨by simp [mul_assoc]⟩
+
+instance is_subgroup_conj (x : G) (H : set G) [is_subgroup H] :
+  is_subgroup (conjugate_set x H) :=
+by rw conjugate_set_eq_image; apply_instance
+
+/-- `dlogn p a` gives the maximum value of `n` such that `p ^ n ∣ a` -/
+def dlogn (p : ℕ) : ℕ → ℕ
+| 0     := 0
+| (a+1) := if h : p > 1 then
+  have (a + 1) / p < a + 1, from nat.div_lt_self dec_trivial h,
+    if p ∣ (a + 1) then dlogn ((a + 1) / p) + 1 else 0
+  else 0

```

Comment:
@robertylewis We have another definition of the largest prime power function here. In your view how does this compare with your version of this definition and theory?

Reply:
This one is certainly more efficient -- mine wasn't written to actually be computed. But the spec proofs are much simpler for mine since they come straight from the more general `nat.find_greatest`. This would have to be extended to integer-valued `n` to use in the p-adics. Otherwise it would be a straightforward replacement, you'd only have to change a handful of lines.

Reply:
My definition can quite easily be generalized
```lean
def nat.find_greatest (P : ℕ → Prop) [decidable_pred P] : ℕ → ℕ
| 0     := 0
| (n+1) := if h : P (n + 1) then n + 1 else nat.find_greatest n
```

Reply:
My point is that defining `dlogn` in terms of `nat.find_greatest`, like I did, is less efficient but shortens the proofs of the  properties that follow.